CuPy + mpi4py container with cotainr (#9)
DeiC-HPC/BitBulldozersLab: mpi4py and cupy container using Cotainr/README.md

# CuPy + mpi4py open source container
- **Keywords:** Container, mpi4py, cupy, open source, MPI, MPICH, osu
- **Date:** 2025-07-24
## Description
This BitBullDozer is a continuation of the [LUMI-C OSU mpi4py benchmarks](https://github.com/DeiC-HPC/cotainr/tree/2025.7.1/examples/LUMI/conda_mpi4py_mpich) example, where we will accomplish the following steps:
- Build a mpi4py container based on the open source containers from the docker build [BitBullDozer](https://github.com/DeiC-HPC/BitBulldozersLab/tree/explore/docker_builds/Docker%20build%20pipeline%20for%204%20different%20container%20approaches%20on%20LUMI)
- Perform the OSU Benchmark and analyse the MPI performance of the mpi4py container
- Build a container with both cupy and mpi4py
- Perform the OSU Benchmark and analyse the MPI performance of the cupy + mpi4py container
- Perform intra-node GPU-to-GPU Benchmark and analyse performance of CuPy.
- Discussion of the settings, tunings, etc. needed for building a container for GPU-GPU MPI communication on LUMI.
- A Jupyter notebook presenting an analysis of the test and benchmark results

> The builds and discussions are found in this README and the benchmark performance are shown in the `lumi_mpi4py_osu_results.ipynb` notebook.

Reviewer: ```suggestion
The builds and discussions are found in this README and the benchmark performance are shown in the `lumi-g_mpi4py_osu_results.ipynb`  and `lumi-c_mpi4py_osu_results.ipynb` notebooks.
```

## Reproduction
This BitBulldozersLab requires the two following **open source** base containers created in the docker build [BitBullDozer](https://github.com/DeiC-HPC/BitBulldozersLab/tree/explore/docker_builds/Docker%20build%20pipeline%20for%204%20different%20container%20approaches%20on%20LUMI)
- `base_image_libcxi_libfabric2000_mpich423.sif`
- `base_image_libcxi_libfabric1220_mpich423.sif`
In this BitBulldozer they are put in

> - `base_images/opensource_base_image_libcxi_libfabric2000_mpich423.sif`

Reviewer: ```suggestion
- `containers/base_images/opensource_base_image_libcxi_libfabric2000_mpich423.sif`
```

> - `base_images/opensource_base_image_libcxi_libfabric1220_mpich423.sif`

Reviewer: ```suggestion
- `containers/base_images/opensource_base_image_libcxi_libfabric1220_mpich423.sif`
```

> With these containers in place, the complete environment containing test method containers and the OSU benchmark can be setup using the `reproduce-environment.sh` script. This script calls `containers/build_container.sh` to build the 8 test method containers which takes approximately 80 minutes.

Reviewer: ```suggestion
With these containers in place, the complete environment containing test method containers and the OSU benchmark can be setup using the `reproduce-environment.sh` script (On Lumi only). This script calls `containers/build_container.sh` to build the 8 test method containers which takes approximately 80 minutes.
```

The GPU intra-node bandwidth performance of the OSU Python benchmark is very fast, where we reach bandwidth speeds upwards of 140GB/s out of the 200GB/s hardware spec. These large further exaggerate the difference between libfabric 1.22 and 2.0 where 1.22 is clearly faster. 

## Issues
When using the `MPIR_CVAR_CH4_OFI_ENABLE_HMEM` flag, we do get an error in using "DMABUF" memory translation
`libfabric:65622:1753693496::core:core:rocr_is_dmabuf_supported():1134<info> DMABUF support: could not open kernel conf file /boot/config-5.14.21-150500.55.49_13.0.56-cray_shasta_c error`

> This can be fixed this by either bind-mounting this kernel file from /boot/ or we can disable dmabuf using `FI_HMEM_ROCR_USE_DMABUF=0`, however it affects performance very little.

Reviewer: ```suggestion
This can be fixed by either bind-mounting this kernel file from /boot/ or we can disable dmabuf using `FI_HMEM_ROCR_USE_DMABUF=0`, however it affects performance very little.
```

In order to use CuPy we required the very newest version for compatibility with Cython 3.1 [4]

The docker to singularity conversion of the base container must be done with `--fix-perms` in order to properly build the containers in this BitBulldozersLab. [5]

> The CuPy recipe requires environment variables in order to compile CuPy correctly, even though the variable are added to the conda environment file (.yml), they are not use and has to be defined manually as cotainr works currently.

Reviewer: What are these environment variables? Or if they are set automatically mention it.

Author: The ones defined in `cupy_mpi4py.yml` under variables: I can add a small comment
